src/caffe/solver.cpp

```diff
@@ -227,13 +210,44 @@ void Solver<Dtype>::Solve(const char* resume_file) {
         }
       }
     }
-
     ComputeUpdateValue();
     net_->Update();
+
+    // Save a snapshot if needed.
+    if (param_.snapshot() && (iter_ + 1) % param_.snapshot() == 0) {
+      Snapshot();
+    }
+  }
+}
+
+template <typename Dtype>
+void Solver<Dtype>::PreSolve() {
```

Comment:
IMO since this method is common setup specific to the `Solver` class itself, it should just be given a new name like `Initialize()` (and then called separately before  `PreSolve()` @ line 236), so that anyone subclassing Solver doesn't need to remember to call it.